### PR TITLE
Candidate for main from GFDL 2020-09-18

### PR DESCRIPTION
**Pull request by adcroft**
This is a PR onto our collective main branch. Note we now use the terminology "main" where we once used "dev/master". Instructions on renaming your branches can be found at https://github.com/NOAA-GFDL/MOM6/wiki/Renaming-the-main-branch-in-MOM6.

This PR has some potentially major-impact changes which are detailed below. Each in isolation made sense but unfortunately we let them stack up so may take some work to sort through - our apologies! 

- A change in defaults for parameters potentially changes your answers if you implicitly rely on a parameter by not setting it explicitly at run-time. If your answers have changed with this PR-to-main then follow the instructions outlined in NOAA-GFDL#1127 to recover your answers.
  - NOAA-GFDL#1127  MOM6: +(*)Changed the defaults for 36 parameters
- Reorganization of some source code. The use of git submodules for optional code in `pkg/` has been replaced by the use of null modules in `config_src/external`. This means the search path for code at compile time must be updated.
  - NOAA-GFDL#1133  Added null version of ocean_BGC
  - NOAA-GFDL#1139  Moved hooks from pkg to external
  - NOAA-GFDL#1160  Move MOM_oda_driver to framework directory
- New version of FMS. It's not a major change but you must use tag 2019.01.03 from here on. There is an new argument in one API that is needed to be forward compatible. We will soon be updating GFDL to using a 2020 tag but that will not require source code change and should be optional.
  - NOAA-GFDL#1141  FMS2020.01 "runtime required" but optional argument to diag_axis_init
  - NOAA-GFDL#1143  This is the minimum required update to use 2020.01 FMS & FMScoupler
  - NOAA-GFDL#1158  Test patch to FMS 2019.01.02 from Rusty
  - NOAA-GFDL#1164  Move FMS tag to 2019.01.03
- Change in diagnostic timing. This will change the values you diagnose for temperatire and salinity. Due to a mistake introduced when closing the remapped diagnostic budgets the call to posts for T/S was moved too early in the algorithm.
  - NOAA-GFDL#1207  Post tracer concentration when synced

I've classified the remaining PRs below which we expect should not affect you adversely.

As usual, please can @jiandewang, @awallcraft, @kshedstrom, and @gustavo-marques review the PR.

Bug fixes:
- NOAA-GFDL#1114  Rotate: fixed-time ALE sponge bugfixes
- NOAA-GFDL#1122  Solution for #1121               (OBC)
- NOAA-GFDL#1128  Fixes a assignment mismatch in MOM_generic_tracers   bug 
- NOAA-GFDL#1129  *Corrected two bugs impacting .testing test cases   bug 
- NOAA-GFDL#1136  +partial fix for #1130
- NOAA-GFDL#1159  +Adding a halo update for tracer reservoirs.
- NOAA-GFDL#1169  Fixing loop bound error in MOM_PressureForce_Montgomery.F90
- NOAA-GFDL#1170  OBC: H-dimensionality fixes
- NOAA-GFDL#1175  (*)Fix code problems that appear in ISOMIP cases
- NOAA-GFDL#1181  (*)Correct the values returned by wave_speeds
- NOAA-GFDL#1183  Fixes in wave structure, solves #1182

Testing:
- NOAA-GFDL#1115  Travis: Arm64 tests; git depth restored; Makefile
- NOAA-GFDL#1118  CodeCov: Set number of reports to 8
- NOAA-GFDL#1119  Codecov: Set base as parent (target) branch
- NOAA-GFDL#1120  Codecov: Set base as parent (target) branch
- NOAA-GFDL#1137  Minor adjustments to .testing/Makefile
- NOAA-GFDL#1138  *Fix frazil restarts (corrects bug from PR#1102)
- NOAA-GFDL#1140  Use FMS tag 2019.01.02 in .testing
- NOAA-GFDL#1144  Testing: Removed environment flag checks
- NOAA-GFDL#1150  Testing: Explicit OpenMP CPU affinity
- NOAA-GFDL#1166  Testing: tc4 installs python-netCDF4 via venv
- NOAA-GFDL#1194  .testing Makefile typo fix
- NOAA-GFDL#1195  Add a codecov upload job to establish coverage changes
- NOAA-GFDL#1197  Autoconf-based build framework
- NOAA-GFDL#1208  Testing: Warn on new diagnostic in regression

Enhancements:
- NOAA-GFDL#1116  MOM6: +Add just_read_params arg to entrain_diffusive_init
- NOAA-GFDL#1123  (*)Improve ePBL when EPBL_ORIGINAL_PE_CALC=False
- NOAA-GFDL#1124  MOM6: +Minor improvements in framework code
- NOAA-GFDL#1125  +(*)Fix conflicting defaults for NKML
- NOAA-GFDL#1126  Neverworld2 and "basin builder"
- NOAA-GFDL#1132  Removing broken link to diag_table documentation.
- NOAA-GFDL#1134  Additions from GMAO (I)
- NOAA-GFDL#1135  +Improve documentation in MOM_parameter_doc files
- NOAA-GFDL#1142  Added diagnostics for partial derivatives of density
- NOAA-GFDL#1147  (*)Corrected halo size in EOS call if VERTEX_SHEAR=T
- NOAA-GFDL#1148  Fixing openmp issues with FMS2020 cpu affinity
- NOAA-GFDL#1152  Removed #ifdef debugging blocks
- NOAA-GFDL#1153  Adding OBC code to slope_x, slope_y
- NOAA-GFDL#1154  +Aligned newlines with module documentation
- NOAA-GFDL#1156  Adds parameterizations of SGS variance to PGF and isopycnal slopes
- NOAA-GFDL#1161  MOM6: +(*)Add option to use time-integrated barotropic continuity
- NOAA-GFDL#1162  MOM6: (*?)Revised ice_shelf_driver.F90 so it compiles
- NOAA-GFDL#1163  Momentum budget terms multiplied by fractional layer-thicknesses
- NOAA-GFDL#1167  OBC: Define global indices when [IJ]s == [IJ]e
- NOAA-GFDL#1171  OBC: Removal of segment zero
- NOAA-GFDL#1172  (*)Improve advective CFL calculation with tiny h
- NOAA-GFDL#1173  Infrastructure calls via framework directory
- NOAA-GFDL#1174  Move call to initialize_segment_data to MOM_state_initialization
- NOAA-GFDL#1176  Dev master candidate NCAR 2020-08-11
- NOAA-GFDL#1178  +Move find_interfaces to MOM_state_initialization
- NOAA-GFDL#1179  New option to use a grid index convention for spatially-varying diagn\u2026
- NOAA-GFDL#1180  Ice shelf grid
- NOAA-GFDL#1184  *+Add halo updates needed with VERTEX_SHEAR=True
- NOAA-GFDL#1186  Fix epipycnal pairing array size declarations
- NOAA-GFDL#1187  Small correction in register_diag_field for 2D diagnostics
- NOAA-GFDL#1188  +Allow position=NORTH_FACE in pass_var calls
- NOAA-GFDL#1189  +Revised interface to write_cputime
- NOAA-GFDL#1190  Updating the default setting for BBL_USE_EOS to match USE_REGRIDDING
- NOAA-GFDL#1192  Always set VARIABLE_BUOYFORCE = True with RESTOREBUOY = True
- NOAA-GFDL#1193  Brankart PGF update
- NOAA-GFDL#1196  *Corrected the test whether to use fluxes%psurf
- NOAA-GFDL#1198  (*)Reduce size of GV%Rlay and improve error handling
- NOAA-GFDL#1199  (*)Correct the use of memory macros for arguments
- NOAA-GFDL#1200  OBC: segment parser refactoring
- NOAA-GFDL#1202  Properly set vertex_shear in set_diffusivity_init
- NOAA-GFDL#1204  +Merge dev/master into dev/gfdl
- NOAA-GFDL#1205  Clean up pass_var for tracer reservoirs.
- NOAA-GFDL#1209  GM-Brankart update from NCAR

### CODE REVIEW

**Comment by kshedstrom** [comment]
I approve this PR.

**Comment by adcroft** [comment]
Following @gustavo-marques realization yesterday, I pushed the corresponding branch "main-candidate-2020-09-18" to MOM6-examples.

**Comment by jiandewang** [comment]
EMC will follow the changing of those default values, works fine in EMC coupled system

**Comment by adcroft** [comment]
@gustavo-marques What's your status on this PR?

**Comment by gustavo-marques** [comment]
@alperaltuntas and I are working on it. Cheyenne was down for maintenance over the last two days. We should be able to finish handling it within the next day or two.

File: config_src/mct_driver/mom_surface_forcing_mct.F90 [hunk 1]
```diff
@@ -1095,7 +1095,7 @@ subroutine surface_forcing_init(Time, G, US, param_file, diag, CS, restore_salt,
     call get_param(param_file, mdl, "USE_NET_FW_ADJUSTMENT_SIGN_BUG", &
                  CS%use_net_FW_adjustment_sign_bug, &
                    "If true, use the wrong sign for the adjustment to "//&
-                   "the net fresh-water.", default=.false.)
+                   "the net fresh-water.", default=.true.)
```
Comment (gustavo-marques) [inline config_src/mct_driver/mom_surface_forcing_mct.F90:1098]: The default for this parameter did not have to be changed. We (NCAR) had already set this to False and, unintentionally, @Hallberg-NOAA set it back to True in NOAA-GFDL#1127

File: config_src/nuopc_driver/mom_surface_forcing_nuopc.F90 [hunk 2]
```diff
@@ -1094,7 +1094,7 @@ subroutine surface_forcing_init(Time, G, US, param_file, diag, CS, restore_salt,
     call get_param(param_file, mdl, "USE_NET_FW_ADJUSTMENT_SIGN_BUG", &
                  CS%use_net_FW_adjustment_sign_bug, &
                    "If true, use the wrong sign for the adjustment to "//&
-                   "the net fresh-water.", default=.false.)
+                   "the net fresh-water.", default=.true.)
```
Comment (gustavo-marques) [inline config_src/nuopc_driver/mom_surface_forcing_nuopc.F90:1097]: Same here. The default should be False.

**Comment by gustavo-marques** [comment]
We were able to recover previous answers after manually setting ``USE_NET_FW_ADJUSTMENT_SIGN_BUG=False``. I suggest that we change the default for this flag in both MCT and NUOPC caps before merging this PR (see above).

**Comment by adcroft** [comment]
@gustavo-marques Thanks for the above. I'll update this branch shortly.

File: config_src/nuopc_driver/mom_surface_forcing_nuopc.F90 (hunk 2 above)
Comment (adcroft) [reply config_src/nuopc_driver/mom_surface_forcing_nuopc.F90]: Addressed in 38710730114452d0111214c72f274a0f8628f451

File: config_src/mct_driver/mom_surface_forcing_mct.F90 (hunk 1 above)
Comment (adcroft) [reply config_src/mct_driver/mom_surface_forcing_mct.F90]: Addressed in 38710730114452d0111214c72f274a0f8628f451

**Comment by gustavo-marques** [comment]
Thanks for making these changes. We now approve this PR.

**Comment by awallcraft** [comment]
I approve this PR.

**Pull request merged**